source/grid/manifold.cc

```diff
@@ -87,6 +87,68 @@ get_new_point (const Quadrature<spacedim> &quad) const
   return p;
 }
 
+template <int dim, int spacedim>
+Point<spacedim>
+Manifold<dim, spacedim>::
+get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
+               const std::vector<double>           &weights) const
+{
+  const double tol = 1e-10;
+  const unsigned int n_points = surrounding_points.size();
+
+  Assert(n_points > 0,
+         ExcMessage("There should be at least one point."));
+
+  Assert(n_points == weights.size(),
+         ExcMessage("There should be as many surrounding points as weights given."));
+
+  Assert(std::abs(std::accumulate(weights.begin(), weights.end(), 0.0)-1.0) < tol,
+         ExcMessage("The weights for the individual points should sum to 1!"));
+
+  // The number of points is small (at most 26 in 3D), therefore try to minize
+  // overhead and use a very simple search of O(N^2) to find the order in which
+  // to loop over points.
+  std::vector<unsigned int> permutation(n_points);
+  std::vector<bool> found(n_points,false);
```

Comment:
Unfortunately no, we can not be entirely sure. This function is public and in principle users can call the function with any number of points and weights. However, any call from within deal.II hands over at most 26 points (the number of points generated by `get_defaults_points_and_weights` in 3D). I can not think of an application that would want to find a new point within a large number of points, but that  does not mean there is no such application. Alternatively, one could also use std::sort with a comparison function that accesses the weights vector (like in the QSorted constructor), but I do not see how to do that without using a lambda function here. In my 3D application the sorting takes about 10% of the total cycles of this function (5% for allocation, 5% for sorting), while 80% are spent in the `get_intermediate_point` function. Not sure about the best way to proceed here.